### DISCUSSION TITLE
Accessing view information from native code.

### DISCUSSION THREAD

**Discussion by Vadoola** [slint-ui/slint#2507] [Q&A] [answered]
I've been looking into Slint a little bit, but haven't dug too deep yet.

I have some crude sample programs, and recently have been trying to get a slightly more substantial test program going. I've been running into an issue that I haven't been able to work around, and was recently reminded to make a post here after seeing [this](https://github.com/slint-ui/slint/discussions/2503) post get made.

I essentially am trying to figure out how to access some information about my view from my native Rust Code.
I have a struct representing remote devices that I want to poll for information.
I have a View for these devices, and they are being shown in a ScrollView.

Since the information being displayed requires monitoring these devices remotely I don't want to be polling them when the devices are not visible in the ScrollView. I couldn't find anyway to get this status from the ScrollView, but could easily calculate it as a property on my DeviceView.

My issue however comes down to how to determine in my Rust code what it's on screen status is in the ScrollView?
My Rust code can access my model of Devices...but I couldn't figure out how to access the DeviceViews. I tried adding an on_screen value in the device struct (not ideal) and updating it from the DeviceView state, but couldn't get that working.
I then tried using a global callback, but like the poster in the above linked Discussion, couldn't figure out how to access the for loop generated DeviceView widgets.

I've created a rough example of what I'm trying to do (using a ListView instead of a ScrollView) that can be seen [here](https://slint-ui.com/releases/1.0.0/editor/?snippet=import+%7B+Button%2C+ListView+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+struct+Device+%7B%0A++++resTime%3A+int%2C%0A++++name%3A+string%2C%0A%7D%0A%0Acomponent+DeviceView+inherits+VerticalLayout+%7B%0A++++in-out+property+%3Cbool%3E+on_screen%3B%0A++++in-out+property%3CDevice%3E+device%3B%0A++++height%3A+120px%3B%0A++++width%3A+120px%3B%0A++++%0A++++%0A++++lbl+%3A%3D+Text+%7B%0A++++++++text%3A+device.name%3B%0A++++%7D%0A++++%0A++++Rectangle+%7B%0A++++++++background%3A+on_screen+%3F+red+%3A+green%3B%0A++++%7D%0A%7D%0A%0Aexport+component+Demo+inherits+Window+%7B%0A++++width%3A+200px%3B%0A++++height%3A+200px%3B%0A++++property+%3C%5BDevice%5D%3E+devices%3A+%5B%0A++++++++%7B%0A++++++++++++resTime%3A+0%2C%0A++++++++++++name%3A+%22Device+1%22%2C%0A++++++++%7D%2C%0A++++++++%7B%0A++++++++++++resTime%3A+0%2C%0A++++++++++++name%3A+%22Device+2%22%2C%0A++++++++%7D%2C%0A++++%5D%3B%0A++++ListView+%7B%0A++++++++for+new_dev%5Bindex%5D+in+root.devices++%3A+DeviceView+%7B%0A++++++++++++device%3A+new_dev%3B%0A++++++++++++%0A++++++++++++%2F%2FJust+to+prove+the+code+is+working+this+changes+to+green+when+half+off+screen%0A++++++++++++%2F%2Finstead+of+fully+off+screen%0A++++++++++++on_screen%3A+%28-parent.viewport-y+%2B+parent.visible-height%2F2%29+%3E+self.y%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A). Essentially right now when the rectangle goes half off the visible area of the ListView it's on_screen property changes from true to false. In this example it changes the color to green, but I need to be able to somehow access that on_screen state from Rust, to determine if I should be polling my devices over the network.

Hope this makes sense.

**Comment by Vadoola** [comment]
So one of the issues I was having trying to get it working with callbacks and the on_screen state in the device struct itself was how to call the callback when the state changes.

I did a bit more digging, and changed my code to change the on_screen state using a pure callback.
```
states [
  on_scr when (...) {
      on_screen: oscr_state(index, true);
  }

  off_scr when (...) {
      on_screen: oscr_state(index, false); 
  }
``` 
I have some rust code to trigger on this callback, updated the appropriate device in the model, and then just return the boolean value passed into it. It's pretty hacky but it's the only thing I've managed to get to compile that seems like it _**could**_ work. However my closure for the oscr_state callback never seems to get called.

**Comment by ogoffart** [comment, marked as answer]
One possible thing is to rely on the lazyness of properties.  The Slint engine will, for example, only query the background property of a rectangle.

So perhaps something like this:
```
Rectangle {
   background: { 
       parent.viewport-y; // just query that property to force a dependence
       SomeGlobal.oscr_state(index);  // call this means we are visible
       return transparent; // or some other color if you want. 
   }
}
```

Note that the oscr_state callback will have to be marked as pure (even if it is conceptually not pure)

**Reply by Vadoola** [reply]
My first quick attempt using this method isn't quite working right, as I don't always get the on_screen state back correctly. However my Rust code is now being called for the callback, and I can update the model with the on screen state.

I'll mark this as the answer for now, as it's the closest I've come to getting this to work, and I'll keep trying to figure out what's wrong in my logic.

Thank you for your help.
